File: android/src/main/java/ly/ulink/reactnative/ULinkBridge.kt
package ly.ulink.reactnative

// ULinkBridge.kt
// Conversion helpers between the native Android SDK types and JS-safe
// Map/List structures that Expo Modules serialises across the bridge.
//
// Rules (global-constraints.md + task-7-brief.md):
//   - config()  MUST force enableDeepLinkIntegration = false regardless of input.
//   - resolvedData() MUST include rawData (serialisation contract with types.ts).
//   - onLog timestamp MUST be a Long (epoch millis) — already a Long in ULinkLogEntry.
//   - sessionState → lowercase string ("IDLE" → "idle", etc.)
//   - JsonElement parameters/metadata/rawData → plain Map/List via recursive conversion.
//   - Do NOT emit iosUrl/androidUrl as resolvedData output keys (they are input-only
//     ULinkParameters fields, not present on ULinkResolvedData).

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import ly.ulink.sdk.models.SessionState
import ly.ulink.sdk.models.SocialMediaTags
import ly.ulink.sdk.models.ULinkConfig
import ly.ulink.sdk.models.ULinkInstallationInfo
import ly.ulink.sdk.models.ULinkParameters
import ly.ulink.sdk.models.ULinkResolvedData
import ly.ulink.sdk.models.ULinkResponse

object ULinkBridge {

    // -------------------------------------------------------------------------
    // config — parse JS Map → ULinkConfig, always forcing enableDeepLinkIntegration=false
    // -------------------------------------------------------------------------

    fun config(map: Map<String, Any?>): ULinkConfig {
        val apiKey = (map["apiKey"] as? String)
            ?: throw IllegalArgumentException("apiKey is required")
        val baseUrl = map["baseUrl"] as? String ?: "https://api.ulink.ly"
        val debug = map["debug"] as? Boolean ?: false
        val persistLastLinkData = map["persistLastLinkData"] as? Boolean ?: true
        val lastLinkTTLSeconds: Long = when (val ttl = map["lastLinkTimeToLiveSeconds"]) {
            is Number -> ttl.toLong()
            else -> 24 * 60 * 60L
        }
        val clearLastLinkOnRead = map["clearLastLinkOnRead"] as? Boolean ?: false
        val redactAll = map["redactAllParametersInLastLink"] as? Boolean ?: false
        @Suppress("UNCHECKED_CAST")
        val redactedKeys: List<String> =
            (map["redactedParameterKeysInLastLink"] as? List<*>)
                ?.filterIsInstance<String>() ?: emptyList()
        val autoCheckDeferredLink = map["autoCheckDeferredLink"] as? Boolean ?: true

        return ULinkConfig(
            apiKey = apiKey,
            baseUrl = baseUrl,
            debug = debug,
            // ALWAYS false — the module owns link delivery via handleDeepLink()
            enableDeepLinkIntegration = false,
            persistLastLinkData = persistLastLinkData,
            lastLinkTimeToLiveSeconds = lastLinkTTLSeconds,
            clearLastLinkOnRead = clearLastLinkOnRead,
            redactAllParametersInLastLink = redactAll,
            redactedParameterKeysInLastLink = redactedKeys,
            autoCheckDeferredLink = autoCheckDeferredLink
        )
    }

    // -------------------------------------------------------------------------
    // parameters — parse JS Map → ULinkParameters
    // -------------------------------------------------------------------------

    fun parameters(map: Map<String, Any?>): ULinkParameters {
        val domain = (map["domain"] as? String)
            ?: throw IllegalArgumentException("domain is required")
        val type = map["type"] as? String ?: "dynamic"

        // socialMediaTags sub-map
        @Suppress("UNCHECKED_CAST")
        val socialTagsMap = map["socialMediaTags"] as? Map<String, Any?>
        val socialTags = socialTagsMap?.let {
            SocialMediaTags(
                ogTitle = it["ogTitle"] as? String,
                ogDescription = it["ogDescription"] as? String,
                ogImage = it["ogImage"] as? String
            )
        }

        // parameters / metadata: JS sends plain Map → convert to JsonElement
        @Suppress("UNCHECKED_CAST")
        val paramsJsonEl = (map["parameters"] as? Map<String, Any?>)
            ?.let { mapToJsonObject(it) }

        @Suppress("UNCHECKED_CAST")
        val metaJsonEl = (map["metadata"] as? Map<String, Any?>)
            ?.let { mapToJsonObject(it) }

        return ULinkParameters(
            type = type,
            domain = domain,
            slug = map["slug"] as? String,
            name = map["name"] as? String,
            iosUrl = map["iosUrl"] as? String,
            androidUrl = map["androidUrl"] as? String,
            iosFallbackUrl = map["iosFallbackUrl"] as? String,
            androidFallbackUrl = map["androidFallbackUrl"] as? String,
            fallbackUrl = map["fallbackUrl"] as? String,
            parameters = paramsJsonEl,
            socialMediaTags = socialTags,
            metadata = metaJsonEl,
            externalId = map["externalId"] as? String
        )
    }

    // -------------------------------------------------------------------------
    // response — ULinkResponse → JS-safe Map
    // -------------------------------------------------------------------------

    fun response(r: ULinkResponse): Map<String, Any?> = mapOf(
        "success" to r.success,
        "url" to r.url,
        "error" to r.error,
        "data" to r.data?.let { jsonObjectToMap(it) }
    )

    // -------------------------------------------------------------------------
    // resolvedData — ULinkResolvedData? → JS-safe Map? (null-safe)
    // Contract with types.ts: must include rawData; must NOT include iosUrl/androidUrl.
    // -------------------------------------------------------------------------

    fun resolvedData(d: ULinkResolvedData?): Map<String, Any?>? {
        d ?: return null
        val socialTagsMap: Map<String, Any?>? = d.socialMediaTags?.let {
            mapOf(
                "ogTitle" to it.ogTitle,
                "ogDescription" to it.ogDescription,
                "ogImage" to it.ogImage
            )
        }
        return mapOf(
            "slug" to d.slug,
            "iosFallbackUrl" to d.iosFallbackUrl,
            "androidFallbackUrl" to d.androidFallbackUrl,
            "fallbackUrl" to d.fallbackUrl,
            "parameters" to d.parameters?.let { jsonElementToAny(it) },
            "socialMediaTags" to socialTagsMap,
            "metadata" to d.metadata?.let { jsonElementToAny(it) },
            "type" to d.type,
            "isDeferred" to d.isDeferred,
            "matchType" to d.matchType,
            // rawData is required per serialisation contract (iOS parity + types.ts)
            "rawData" to d.rawData?.let { jsonObjectToMap(it) }
        )
    }

    // -------------------------------------------------------------------------
    // installationInfo — ULinkInstallationInfo? → JS-safe Map?
    // -------------------------------------------------------------------------

    fun installationInfo(info: ULinkInstallationInfo?): Map<String, Any?>? {
        info ?: return null
        return mapOf(
            "installationId" to info.installationId,
            "isReinstall" to info.isReinstall,
            "previousInstallationId" to info.previousInstallationId,
            "reinstallDetectedAt" to info.reinstallDetectedAt,
            "persistentDeviceId" to info.persistentDeviceId
        )
    }

    // -------------------------------------------------------------------------
    // sessionState — SessionState enum → lowercase string (matches types.ts enum values)
    // -------------------------------------------------------------------------

    fun sessionState(s: SessionState): String = when (s) {
        SessionState.IDLE -> "idle"
        SessionState.INITIALIZING -> "initializing"
        SessionState.ACTIVE -> "active"
        SessionState.ENDING -> "ending"
        SessionState.FAILED -> "failed"
    }

    // -------------------------------------------------------------------------
    // Private: recursive JsonElement → plain JVM types (Map/List/primitives)
    // The Expo bridge can serialise these to JS without extra work.
    // -------------------------------------------------------------------------

    internal fun jsonElementToAny(el: JsonElement): Any? = when {
        el is JsonNull -> null
        el is JsonArray -> el.map { jsonElementToAny(it) }
        el is JsonObject -> jsonObjectToMap(el)
        el is JsonPrimitive && el.isString -> el.content
        el is JsonPrimitive && el.content == "true" -> true
        el is JsonPrimitive && el.content == "false" -> false
        el is JsonPrimitive && el.content.toLongOrNull() != null -> el.content.toLong()
        el is JsonPrimitive && el.content.toDoubleOrNull() != null -> el.content.toDouble()
        else -> (el as? JsonPrimitive)?.content
    }

    internal fun jsonObjectToMap(obj: JsonObject): Map<String, Any?> =
        obj.entries.associate { (k, v) -> k to jsonElementToAny(v) }

    // Convert a plain Map<String, Any?> to a JsonObject for ULinkParameters fields
    private fun mapToJsonObject(map: Map<String, Any?>): JsonObject = buildJsonObject {
        map.forEach { (key, value) ->
            when (value) {
                null -> put(key, JsonNull)
                is String -> put(key, JsonPrimitive(value))
                is Boolean -> put(key, JsonPrimitive(value))
                is Number -> put(key, JsonPrimitive(value.toDouble()))
                else -> put(key, JsonPrimitive(value.toString()))
            }
        }
    }
}

File: ios/ULinkBridge.swift
// ULinkBridge.swift
// Helper types and the ULinkPendingQueue used by ULinkReactNativeModule.
//
// ULinkPendingQueue:
//   - Queues all JS-originated calls that arrive before initialize() resolves.
//   - markReady() drains the queue on the actor's executor (serialised, no races).
//   - Every pending item carries a typed continuation so the original Promise
//     settles correctly once the SDK is up.
//
// Note: link-delivery (AppDelegate universal-link / URL-scheme forwarding) is
// handled by a separate Expo AppDelegate subscriber (Task 6).  This file only
// owns the method/event surface.

import Foundation
import ULinkSDK
import Combine

// ---------------------------------------------------------------------------
// MARK: - SessionState → string
// ---------------------------------------------------------------------------

func sessionStateString(_ state: SessionState) -> String {
    switch state {
    case .idle:         return "idle"
    case .initializing: return "initializing"
    case .active:       return "active"
    case .ending:       return "ending"
    case .failed:       return "failed"
    }
}

// ---------------------------------------------------------------------------
// MARK: - Serialisation helpers
// ---------------------------------------------------------------------------

/// ULinkResolvedData → JS-safe [String: Any?]
func resolvedDataToMap(_ data: ULinkResolvedData) -> [String: Any?] {
    var socialMediaTagsMap: [String: Any?]? = nil
    if let tags = data.socialMediaTags {
        socialMediaTagsMap = [
            "ogTitle":       tags.ogTitle,
            "ogDescription": tags.ogDescription,
            "ogImage":       tags.ogImage,
        ]
    }

    return [
        "slug":                 data.slug,
        "iosFallbackUrl":       data.iosFallbackUrl,
        "androidFallbackUrl":   data.androidFallbackUrl,
        "fallbackUrl":          data.fallbackUrl,
        "parameters":           data.parameters,
        "socialMediaTags":      socialMediaTagsMap,
        "metadata":             data.metadata,
        "rawData":              data.rawData,
        "type":                 data.type,
        "isDeferred":           data.isDeferred,
        "matchType":            data.matchType,
    ]
}

/// ULinkResponse → JS-safe [String: Any?]
func responseToMap(_ resp: ULinkResponse) -> [String: Any?] {
    return [
        "success": resp.success,
        "url":     resp.url,
        "error":   resp.error,
        "data":    resp.data,
    ]
}

/// ULinkInstallationInfo (struct, not class) → JS-safe [String: Any?]
func installationInfoToMap(_ info: ULinkInstallationInfo) -> [String: Any?] {
    return [
        "installationId":        info.installationId,
        "isReinstall":           info.isReinstall,
        "previousInstallationId": info.previousInstallationId,
        "reinstallDetectedAt":   info.reinstallDetectedAt,
        "persistentDeviceId":    info.persistentDeviceId,
    ]
}

/// Parses a JS config dict into ULinkConfig.
/// enableDeepLinkIntegration is always forced to false; link delivery is Task 6.
func parseConfig(_ map: [String: Any]) throws -> ULinkConfig {
    guard let apiKey = map["apiKey"] as? String, !apiKey.isEmpty else {
        throw NSError(
            domain: "ULinkReactNative",
            code: 1,
            userInfo: [NSLocalizedDescriptionKey: "apiKey is required"]
        )
    }
    let baseUrl               = map["baseUrl"] as? String ?? "https://api.ulink.ly"
    let debug                 = map["debug"] as? Bool ?? false
    let persistLastLinkData   = map["persistLastLinkData"] as? Bool ?? false
    var lastLinkTTL: TimeInterval? = nil
    if let ms = map["lastLinkTimeToLiveSeconds"] as? Double {
        lastLinkTTL = ms          // JS passes seconds (see types.ts field name)
    }
    let clearLastLinkOnRead               = map["clearLastLinkOnRead"] as? Bool ?? false
    let redactAll                         = map["redactAllParametersInLastLink"] as? Bool ?? false
    let redactedKeys: [String]            = map["redactedParameterKeysInLastLink"] as? [String] ?? []
    let autoCheckDeferredLink             = map["autoCheckDeferredLink"] as? Bool ?? true

    return ULinkConfig(
        apiKey: apiKey,
        baseUrl: baseUrl,
        debug: debug,
        enableDeepLinkIntegration: false,   // module owns link delivery (Task 6)
        persistLastLinkData: persistLastLinkData,
        lastLinkTimeToLive: lastLinkTTL,
        clearLastLinkOnRead: clearLastLinkOnRead,
        redactAllParametersInLastLink: redactAll,
        redactedParameterKeysInLastLink: redactedKeys,
        autoCheckDeferredLink: autoCheckDeferredLink
    )
}

/// Parses a JS parameters dict into ULinkParameters.
func parseParameters(_ map: [String: Any]) throws -> ULinkParameters {
    guard let domain = map["domain"] as? String, !domain.isEmpty else {
        throw NSError(
            domain: "ULinkReactNative",
            code: 2,
            userInfo: [NSLocalizedDescriptionKey: "domain is required"]
        )
    }
    let type                 = map["type"] as? String ?? "dynamic"
    let slug                 = map["slug"] as? String
    let name                 = map["name"] as? String
    let iosFallbackUrl       = map["iosFallbackUrl"] as? String
    let androidFallbackUrl   = map["androidFallbackUrl"] as? String
    let fallbackUrl          = map["fallbackUrl"] as? String
    let iosUrl               = map["iosUrl"] as? String
    let androidUrl           = map["androidUrl"] as? String
    let parameters           = map["parameters"] as? [String: Any]
    let metadata             = map["metadata"] as? [String: Any]
    let externalId           = map["externalId"] as? String

    var socialTags: SocialMediaTags? = nil
    if let socialMap = map["socialMediaTags"] as? [String: Any] {
        socialTags = SocialMediaTags(
            ogTitle:       socialMap["ogTitle"] as? String,
            ogDescription: socialMap["ogDescription"] as? String,
            ogImage:       socialMap["ogImage"] as? String
        )
    }

    if type == "unified" {
        return ULinkParameters.unified(
            domain:         domain,
            slug:           slug,
            name:           name,
            iosUrl:         iosUrl ?? "",
            androidUrl:     androidUrl ?? "",
            fallbackUrl:    fallbackUrl ?? "",
            parameters:     parameters,
            socialMediaTags: socialTags,
            metadata:       metadata,
            externalId:     externalId
        )
    } else {
        return ULinkParameters.dynamic(
            domain:              domain,
            slug:                slug,
            name:                name,
            iosFallbackUrl:      iosFallbackUrl,
            androidFallbackUrl:  androidFallbackUrl,
            fallbackUrl:         fallbackUrl,
            parameters:          parameters,
            socialMediaTags:     socialTags,
            metadata:            metadata,
            externalId:          externalId
        )
    }
}

// ---------------------------------------------------------------------------
// MARK: - Incoming-link buffer (used by the AppDelegate subscriber)
// ---------------------------------------------------------------------------

/// Singleton actor that buffers URLs delivered by the AppDelegate subscriber
/// (universal links and custom-scheme URLs) until BOTH conditions are true:
///   1. The SDK has been initialised (`setReady` called from `initialize()`).
///   2. A JS listener is attached (`setObserving` called from `OnStartObserving`).
///
/// This two-gate approach prevents the cold-start race where a link is emitted
/// before JS has called `ULink.onDynamicLink(cb)` / `ULink.onUnifiedLink(cb)`.
///
/// Usage pattern:
///   AppDelegate subscriber:
///     Task { await ULinkIncomingLinkBuffer.shared.buffer(url) }
///
///   Module (inside initialize(), after subscribeStreams + markReady):
///     await ULinkIncomingLinkBuffer.shared.setReady(sdk)
///
///   Module definition():
///     OnStartObserving { Task { await ULinkIncomingLinkBuffer.shared.setObserving() } }
///
///   Module didDispose():
///     Task { await ULinkIncomingLinkBuffer.shared.reset() }
actor ULinkIncomingLinkBuffer {

    static let shared = ULinkIncomingLinkBuffer()
    private init() {}

    private var sdk: ULink? = nil
    private var observing: Bool = false
    private var buffered: [URL] = []

    /// Called by the module after `ULink.initialize()` succeeds and Combine
    /// streams are live.  Stores the SDK reference and flushes if a JS listener
    /// is already attached.
    func setReady(_ sdk: ULink) async {
        self.sdk = sdk
        await flushIfPossible()
    }

    /// Called when the first JS listener attaches (via `OnStartObserving`).
    /// Marks the module as observed and flushes if the SDK is also ready.
    func setObserving() async {
        observing = true
        await flushIfPossible()
    }

    /// Called by the AppDelegate subscriber with each incoming URL.
    /// Emits immediately if both gates are open; otherwise buffers.
    func buffer(_ url: URL) async {
        if let sdk = sdk, observing {
            do {
                try await sdk.handleDeepLinkAsync(url: url)
            } catch {
                // Swallow: no JS Promise to reject for AppDelegate-sourced URLs.
            }
        } else {
            buffered.append(url)
        }
    }

    /// Drains all buffered URLs through the SDK when both gates are open.
    private func flushIfPossible() async {
        guard let sdk = sdk, observing else { return }
        let pending = buffered
        buffered.removeAll()
        for url in pending {
            do {
                try await sdk.handleDeepLinkAsync(url: url)
            } catch {
                // Swallow: background delivery; no Promise to reject.
            }
        }
    }

    /// Resets the buffer on module dispose so a stale SDK reference is never
    /// used after the module is torn down.  `observing` is intentionally kept
    /// so that a re-init + already-attached listener still flushes correctly.
    func reset() async {
        sdk = nil
        buffered.removeAll()
    }
}

// ---------------------------------------------------------------------------
// MARK: - Pending-call infrastructure
// ---------------------------------------------------------------------------

/// A single queued call waiting for initialisation.
/// Each case captures a typed continuation so the original Promise is resolved
/// (or rejected) after markReady() drains the queue.
enum PendingCall {
    case initialize(
        config: [String: Any],
        resolve: (Void) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case createLink(
        params: [String: Any],
        resolve: ([String: Any?]) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case resolveLink(
        url: String,
        resolve: ([String: Any?]) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case processULink(
        url: String,
        resolve: ([String: Any?]?) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case checkDeferredLink(
        resolve: (Void) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case getInitialDeepLink(
        resolve: ([String: Any?]?) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case getInitialUri(
        resolve: (String?) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case setInitialUri(
        uri: String,
        resolve: (Void) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case getLastLinkData(
        resolve: ([String: Any?]?) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case getInstallationId(
        resolve: (String?) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case getInstallationInfo(
        resolve: ([String: Any?]?) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case isReinstall(
        resolve: (Bool) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case getCurrentSessionId(
        resolve: (String?) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case hasActiveSession(
        resolve: (Bool) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case getSessionState(
        resolve: (String) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case endSession(
        resolve: (Void) -> Void,
        reject:  (String, String, Error?) -> Void
    )
    case dispose(
        resolve: (Void) -> Void,
        reject:  (String, String, Error?) -> Void
    )
}

// ---------------------------------------------------------------------------
// MARK: - ULinkPendingQueue actor
// ---------------------------------------------------------------------------

/// Thread-safe queue that buffers JS calls arriving before `initialize()` resolves.
///
/// Usage pattern:
///   1. Module methods call `queue.enqueue(.someCase(...))` when the SDK is not yet ready.
///   2. After `ULink.initialize(config:)` succeeds the module calls `queue.markReady(ulink:)`.
///   3. `markReady` drains the queue, executing each pending call against the live SDK.
actor ULinkPendingQueue {

    private var pending: [PendingCall] = []
    private var ready = false
    private var ulink: ULink? = nil

    /// Called by the module after `ULink.initialize(config:)` returns successfully.
    func markReady(_ sdk: ULink, module: ULinkReactNativeModule) async {
        self.ulink = sdk
        self.ready = true
        let calls = pending
        pending.removeAll()
        for call in calls {
            await execute(call, module: module)
        }
    }

    /// Enqueue a call.  If already ready, execute immediately (no queue).
    func enqueue(_ call: PendingCall, module: ULinkReactNativeModule) async {
        if ready, let sdk = ulink {
            await execute(call, module: module)
        } else {
            pending.append(call)
        }
    }

    // swiftlint:disable:next cyclomatic_complexity function_body_length
    private func execute(_ call: PendingCall, module: ULinkReactNativeModule) async {
        guard let sdk = ulink else { return }

        switch call {

        case .initialize(_, let resolve, _):
            // Already initialised — idempotent, resolve immediately.
            resolve(())

        case .createLink(let params, let resolve, let reject):
            do {
                let p = try parseParameters(params)
                let resp = try await sdk.createLink(parameters: p)
                resolve(responseToMap(resp) as [String: Any?])
            } catch {
                reject("CREATE_LINK_ERROR", error.localizedDescription, error)
            }

        case .resolveLink(let url, let resolve, let reject):
            do {
                let resp = try await sdk.resolveLink(url: url)
                resolve(responseToMap(resp) as [String: Any?])
            } catch {
                reject("RESOLVE_LINK_ERROR", error.localizedDescription, error)
            }

        case .processULink(let url, let resolve, let reject):
            guard let linkUrl = URL(string: url) else {
                reject("INVALID_URL", "Invalid URL: \(url)", nil)
                return
            }
            do {
                let data = try await sdk.processULinkUrlThrowing(linkUrl)
                resolve(data.map { resolvedDataToMap($0) as [String: Any?] })
            } catch {
                reject("PROCESS_ULINK_ERROR", error.localizedDescription, error)
            }

        case .checkDeferredLink(let resolve, let reject):
            do {
                try await sdk.checkDeferredLinkAsync()
                resolve(())
            } catch {
                reject("DEFERRED_LINK_ERROR", error.localizedDescription, error)
            }

        case .getInitialDeepLink(let resolve, _):
            let data = await sdk.getInitialDeepLink()
            resolve(data.map { resolvedDataToMap($0) as [String: Any?] })

        case .getInitialUri(let resolve, _):
            resolve(sdk.getInitialUrl()?.absoluteString)

        case .setInitialUri(let uri, let resolve, let reject):
            guard let url = URL(string: uri) else {
                reject("INVALID_URL", "Invalid URI: \(uri)", nil)
                return
            }
            sdk.setInitialUrl(url)
            resolve(())

        case .getLastLinkData(let resolve, _):
            let data = sdk.getLastLinkData()
            resolve(data.map { resolvedDataToMap($0) as [String: Any?] })

        case .getInstallationId(let resolve, _):
            resolve(sdk.getInstallationId())

        case .getInstallationInfo(let resolve, _):
            let info = sdk.getInstallationInfo()
            resolve(info.map { installationInfoToMap($0) as [String: Any?] })

        case .isReinstall(let resolve, _):
            resolve(sdk.isReinstall())

        case .getCurrentSessionId(let resolve, _):
            resolve(sdk.getCurrentSessionId())

        case .hasActiveSession(let resolve, _):
            resolve(sdk.hasActiveSession())

        case .getSessionState(let resolve, _):
            resolve(sessionStateString(sdk.getSessionState()))

        case .endSession(let resolve, _):
            _ = await sdk.endSession()
            resolve(())

        case .dispose(let resolve, _):
            sdk.dispose()
            module.didDispose()
            resolve(())
        }
    }
}
